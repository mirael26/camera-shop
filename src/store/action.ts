import { IChangeAddingToCartItem, ICloseModal, ILoadCurrentProduct, ILoadProducts, ILoadPromo, ILoadReviews, IOpenModal, IRedirect } from '../types/actions.type';
import { TModal } from '../types/app.type';
import { IProduct, IPromo, IReview } from '../types/data.type';

export const ActionType = {
  LoadPromo: 'promo/loadPromo',
  LoadProducts: 'cameras/loadProducts',
  LoadCurrentProduct: 'cameras/loadCurrentProduct',
  LoadReviews: 'cameras/loadReviews',
  OpenModal: 'state/openModal',
  CloseModal: 'state/closeModal',
  ChangeAddingToCartItem: 'state/changeAddingToCartItem',
  Redirect: 'state/redirect',
} as const;

export const ActionCreator = {
  LoadPromo: (promo: IPromo): ILoadPromo => ({
    type: ActionType.LoadPromo,
    payload: promo,
  }),
  LoadProducts: (products: Array<IProduct>): ILoadProducts => ({
    type: ActionType.LoadProducts,
    payload: products,
  }),
  LoadCurrentProduct: (product: IProduct): ILoadCurrentProduct => ({
    type: ActionType.LoadCurrentProduct,
    payload: product,
  }),
  LoadReviews: (reviews: Array<IReview> | null): ILoadReviews => ({
    type: ActionType.LoadReviews,
    payload: reviews,
  }),
  OpenModal: (modal: TModal): IOpenModal => ({
    type: ActionType.OpenModal,
    payload: modal,
  }),
  CloseModal: (modal: TModal): ICloseModal => ({
    type: ActionType.CloseModal,
    payload: modal,
  }),
  ChangeAddingToCartItem: (item: IProduct | null): IChangeAddingToCartItem => ({
    type: ActionType.ChangeAddingToCartItem,
    payload: item,
  }),
  Redirect: (path: string | null): IRedirect => ({
    type: ActionType.Redirect,
    payload: path,
  }),
};
