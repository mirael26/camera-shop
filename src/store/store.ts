import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { dataReducer } from './reducers/data-reducer';
import { stateReducer } from './reducers/state-reducer';

const rootReducer = combineReducers({data: dataReducer, state: stateReducer});

const store = configureStore({
  reducer: rootReducer,
  devTools: true,
});

export type TRootReducer = ReturnType<typeof store.getState>;
export type TAppDispatch = typeof store.dispatch;

export default store;
