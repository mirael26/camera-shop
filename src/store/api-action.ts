import axios from 'axios';
import { ApiUrl } from '../consts';
import { IPromo } from '../types/data.type';
import { ActionCreator } from './action';
import { TAppDispatch } from './store';

const URL = 'https://camera-shop.accelerator.pages.academy';

export const loadPromo = () => (dispatch: TAppDispatch) => {
  axios
    .get(`${URL}${ApiUrl.Promo}`)
    .then((response) => dispatch(ActionCreator.LoadPromo(response.data as IPromo)))
    .catch((error) => {
      throw(error);
    });
};
