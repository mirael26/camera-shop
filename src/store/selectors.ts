import { createSelector } from 'reselect';
import { TRootReducer } from './store';

export const getAllProducts = (state: TRootReducer) => state.data.products;
export const getAllProductsCount = (state: TRootReducer) => state.data.products?.length;
export const getFilteredProductsCount = (state: TRootReducer) => state.data.filteredProducts?.length;
export const getFilteredExcludingPriceProducts = (state: TRootReducer) => state.data.filteredExcludingPriceProducts;
export const getDisplayedProducts = (state: TRootReducer) => state.data.displayedProducts;
export const getCurrentProduct = (state: TRootReducer) => state.data.currentProduct;
export const getSimilarProducts = (state: TRootReducer) => state.data.similarProducts;
export const getReviews = (state: TRootReducer) => state.data.reviews;
export const getPromo = (state: TRootReducer) => state.data.promo;

export const getProductsInCart = (state: TRootReducer) => state.cart.productsInCart;
export const getAddedToCartItem = (state: TRootReducer) => state.cart.addedToCartItem;
export const getDeletedFromCartItem = (state: TRootReducer) => state.cart.deletedFromCartItem;
export const getPromocode = (state: TRootReducer) => state.cart.promocode;
export const getPromocodeConfirmed = (state: TRootReducer) => state.cart.isPromocodeConfirmed;
export const getDiscount = (state: TRootReducer) => state.cart.discount;

export const getModalStatus = (state: TRootReducer) => state.view.isModalOpen;
export const getActiveModal = (state: TRootReducer) => state.view.activeModal;
export const getRedirectPath = (state: TRootReducer) => state.view.redirect;
export const getProductsLoadingStatus = (state: TRootReducer) => state.view.productsIsLoading;

export const getAllProductsAsc = createSelector(
  getAllProducts,
  (products) => {
    const productsAsc = products?.slice();
    productsAsc?.sort((a, b) => a.price - b.price);
    return productsAsc;
  }
);

export const getFilteredExcludingPriceProductsAsc = createSelector(
  getFilteredExcludingPriceProducts,
  (filteredExcludingPriceProducts) => {
    const filteredExcludingPriceProductsAsc = filteredExcludingPriceProducts?.slice();
    filteredExcludingPriceProductsAsc?.sort((a, b) => a.price - b.price);
    return filteredExcludingPriceProductsAsc;
  }
);

export const getProductsInCartCount = createSelector(
  getProductsInCart,
  (products) => {
    if (products.length === 0) {
      return 0;
    }
    const count = products.reduce((accumulator, product) => accumulator + product.countInCart, 0);
    return count;
  }
);
