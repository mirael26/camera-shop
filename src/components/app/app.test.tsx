import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createReduxStore } from '../../store/store';
import { renderWithReduxAndRouter } from '../../test/helpers/render-with-redux-and-router';
import App from './app';

jest.mock('../redirect/redirect', () => 'Redirect');
jest.mock('./app-router/app-router', () => 'AppRouter');

test('Render App', () => {
  const { asFragment } = renderWithReduxAndRouter(<App/>);
  expect(asFragment()).toMatchSnapshot();
});
