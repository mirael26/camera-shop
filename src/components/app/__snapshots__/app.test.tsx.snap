// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App 1`] = `
<DocumentFragment>
  <scrolltotop />
  <redirect />
  <approuter />
</DocumentFragment>
`;
